packages/react-native-reanimated/src/platform-specific/RNRenderer.ts

```diff
@@ -1,4 +1,44 @@
+/* eslint-disable camelcase */
 // eslint-disable-next-line @typescript-eslint/ban-ts-comment
 // @ts-nocheck
 'use strict';
-export { default as RNRenderer } from 'react-native/Libraries/Renderer/shims/ReactNative';
+
+import { isFabric } from '../PlatformChecker';
+
+function findHostInstanceFastPath(ref: React.Component) {
+  if (ref.__internalInstanceHandle && ref.__nativeTag && ref._viewConfig) {
+    return ref;
+  }
+  return undefined;
+}
+
+function resolveFindHostInstance_DEPRECATED() {
+  if (findHostInstance_DEPRECATED !== undefined) {
+    return;
+  }
+  if (isFabric()) {
+    try {
+      findHostInstance_DEPRECATED =
+        require('react-native/Libraries/Renderer/shims/ReactFabric').findHostInstance_DEPRECATED;
+    } catch (e) {
+      findHostInstance_DEPRECATED = (_ref: unknown) => null;
+    }
+  } else {
+    findHostInstance_DEPRECATED =
+      require('react-native/Libraries/Renderer/shims/ReactNative').findHostInstance_DEPRECATED;
+  }
+}
+
+let findHostInstance_DEPRECATED: (ref: unknown) => void;
+export function findHostInstance(component: React.Component): unknown {
+  // Fast path for native refs
+  const hostInstance = findHostInstanceFastPath(component._componentRef);
+  if (hostInstance !== undefined) {
+    return hostInstance;
+  }
+
+  resolveFindHostInstance_DEPRECATED();
```

Comment:
Can we call it in top-level instead or do some dynamic shenanigans instead? Or is it necessary to try resolve it on each `findHostInstance` call?

Reply:
The `resolveFindHostInstance_DEPRECATED` method uses the `isFabric` method, which has caused some race conditions in the past when used at the top level. Sometimes, the JS code was executed before the native part had a chance to inject information about the architecture type.